I've read that while/else contructs run the `else` block only if the loop is broken out of with a `break`. Now I'm watching a Python guy on YouTube and he's saying the exact opposite is true for for/else blocks. Lets try it out and maybe clear up this confusion..

In [1]:
# I started watching this video for zip/map comparisons
a = list(range(10, 101, 10))
b = list(range(1000, 10_001, 1000))
c = list(map(lambda x, y: y - x, a, b))
print(c)

[990, 1980, 2970, 3960, 4950, 5940, 6930, 7920, 8910, 9900]


In [2]:
# Test that while/else runs the else block after a `break`
t = 0
while t < len(c):
    x = c[t]
    if x > 5000:
        break
    t += 1
else:
    print('Hello from while/else break test')

# Test that for/else runs the else block after a `break`
for x in c:
    if x > 5000:
        break
else:
    print('Hello from for/else break test')

So neither of these else clauses run. The video might be for Python 2.7 though, ~~and I'm unable to test with that~~ which doesn't matter because, after testing at [repl.it](https://repl.it/@tener/IrresponsibleLightskyblueBengaltiger), it isn't correct for Python 2.7 either. The loops run fine. `x` is correctly 5940 in each snippet, but it never says 'Hello.' :-(

Trying a different way, we'll let the loops run to completion by changing the `break`s to `pass`es

In [3]:
# Test that while/else runs the else block after a `break`
t = 0
while t < len(c):
    x = c[t]
    if x > 5000:
        pass
    t += 1
else:
    print('Hello from while/else pass test')

# Test that for/else runs the else block after a `break`
for x in c:
    if x > 5000:
        pass
else:
    print('Hello from for/else pass test')

Hello from while/else pass test
Hello from for/else pass test


###### `t = 0`
If the loop runs all the way through without running into a `break` statement, the `else` block runs.

###### `t = 10`
If the loop doesn't run at all because its start condition was never satisfied, the `else` block runs.

###### `break`
I think the use here is `break`, but in a [non-obvious way](http://python-notes.curiousefficiency.org/en/latest/python_concepts/break_else.html). If the loop is `break`ed out of, the `else` statement does **not** run. From this test, it seems the only way to make the `else` block *not* run is to `break` out of the `for` or `while` block.

### Interesting Note

Nick Coghlan, from the link in that last comment, also points out this little tidbit to try and help explain this behaviour.

In [13]:
a = list(range(10))

try:
    a[10] = 1
except Exception as e:
    print('exception raised')
else:
    print('no exception raised')

print('...rest of program...')

exception raised
...rest of program...


This makes much more sense to me, but now I'm wondering why not give the for/else and while/else constructs a `finally:` block as well...